src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp

```diff
@@ -0,0 +1,112 @@
+    } else {
+      return false;
+    }
+  }();
+
+  const bool comoving_decreasing_slower_than_char_speeds =
+      not(crossing_time_info.t_char_speed > 0.0 &&
+          crossing_time_info.t_comoving_char_speed > 0.0 &&
+          update_args.min_comoving_char_speed > 0.0 &&
+          update_args.min_comoving_char_speed /
+                  crossing_time_info.t_comoving_char_speed >
+              update_args.min_char_speed / crossing_time_info.t_char_speed);
+
+  if (char_speed_is_in_danger) {
+    constexpr double min_char_speed_increase_factor = 1.01;
+    if (info->target_char_speed < min_char_speed) {
+      // We are already in state 1, and we are in danger.
+      // But TargetSpeed is less than CharSpeedMin, so we don't want
+      // to continue to push CharSpeedMin downward.  Instead, increase
+      // TargetSpeed to be above CharSpeedMin.
+      // We don't do this if DeltaRexIsInDanger, because for DeltaRex
+      // in danger, we might need to drive CharSpeed lower.
+      info->target_char_speed = min_char_speed * min_char_speed_increase_factor;
+    }
+    info->suggested_time_scale = crossing_time_info.t_char_speed;
+  } else if (delta_radius_is_in_danger) {
+    constexpr double target_speed_decrease_factor = 0.125;
+    constexpr double time_tolerance_for_delta_r =
+        0.25 * time_tolerance_for_delta_r_in_danger;
+    if ((crossing_time_info.t_char_speed > 0.0 and
+         crossing_time_info.t_delta_radius >
+             info->damping_time * time_tolerance_for_delta_r) or
+        update_args.min_comoving_char_speed < 0.0) {
+      info->discontinuous_change_has_occurred = true;
+      info->target_char_speed = min_char_speed * target_speed_decrease_factor;
+      info->suggested_time_scale =
+          std::min(info->damping_time, crossing_time_info.t_delta_radius);
+    } else {
+      info->discontinuous_change_has_occurred = true;
+      info->state = SizeControlLabel::DeltaR;
+      info->suggested_time_scale = crossing_time_info.t_delta_radius;
+      // TODO: PreferState3OverState2 goes here.
+    }
+  } else if (update_args.min_comoving_char_speed > 0.0 and
+             update_args.min_char_speed > 0.0 and
+             (crossing_time_info.t_comoving_char_speed == 0.0 or
+              (crossing_time_info.t_comoving_char_speed >
+                   5.0 * info->damping_time and
+               comoving_decreasing_slower_than_char_speeds)) and
+             (update_args.min_char_speed >= info->target_char_speed or
+              min_comoving_char_speed > min_char_speed)) {
+    info->discontinuous_change_has_occurred = true;
+    info->state = SizeControlLabel::DeltaR;
+    // TODO: PreferState3OverState2 goes here.
+  }
+  // If no 'if's are encountered above, then all the info parameters stay
+  // the same as they were.
+}
+
+double AhSpeed::control_signal(
+    const SizeControlInfo& info,
+    const SizeControlStateControlSignalArgs& control_signal_args) const {
+  const double Y00 = sqrt(0.25 / M_PI);
```

Comment:
`constexpr`?

Reply:
constexpr didn't work; it gave an error about sqrt not being contexpr.

Reply:
https://stackoverflow.com/questions/8622256/in-c11-is-sqrt-defined-as-constexpr